### ISSUE TITLE
Tracking: implementation of RFC 0146 meta.categories

### ISSUE THREAD

**Issue by Aleksanaa** [NixOS/nixpkgs#360727]
Since Anderson Torres said they don't have time recently, we probably should start this work sooner rather than later.

### RFC:

https://github.com/NixOS/rfcs/blob/master/rfcs/0146-meta-categories.md
https://github.com/NixOS/rfcs/pull/146

### Prior attempts:

- https://github.com/NixOS/nixpkgs/pull/230439
- https://github.com/NixOS/nixpkgs/pull/325357

### Main tasks

- [x] Establish an initial categorization team
- [ ] Discuss main directions and reference targets
- [ ] Decide an initial list of categories
- [ ] Implement meta.categories field
- [ ] Motivate people to label as many categories as possible
- [ ] Make `nix search` and search.nixos.org support categories?

-----

CC @AndersonTorres @7c6f434c @natsukium @fgaz @GetPsyched @Lyndeno @pyrotelekinetic

**Comment by tomodachi94** [comment]
I think it would be good to add the field before deciding which categories will initially be added, to avoid a major technical piece from being delayed. I think it would also be reasonable to add the categories one-by-one as the team starts categorizing things (thus, no categories will be added that aren't used).

**Comment by 7c6f434c** [comment]
My personal opinion is that adding a field should reasonably happen after having at least one category defined, and adding at least one category defined should reasonably happen after writing down at least a general direction of how the base categories. This direction design should have at least a majority support among people having actively confirmed the participation in the categorisation group.

**Comment by tomodachi94** [comment]
@NixOS/categorization: would a Matrix channel be helpful for coordination?

**Comment by tomodachi94** [comment]
Created at https://matrix.to/#/#nixpkgs-categorization:matrix.org. I invited all of the team members whose Matrix handles I could find.

**Comment by l0b0** [comment]
Possible plan + surrounding details, as discussed [on Matrix](https://matrix.to/#/#nixpkgs-categorization:matrix.org):

# Goals

## Must haves

- [ ] Make it easy for search.nixos.org to implement filtering based on
      categories. This should follow from the next item.
- [ ] Make it easy to filter packages by category in the Nix REPL. Filtering by
      license looks like this:

    ```nix
    let
      targetLicense = lib.licenses.mit;
    in
    lib.filter (
      package:
      let
        result = lib.tryEval (
          package ? meta
          && package.meta ? license
          && (
            if lib.isList package.meta.license then
              lib.elem targetLicense package.meta.license
            else
              targetLicense == package.meta.license
          )
        );
      in
      result.success && result.value
    ) (lib.attrValues legacyPackages.x86_64-linux)
    ```

    Filtering by category should look something like this:

    ```nix
    lib.filter (
      package:
      let
        result = lib.tryEval (
          package ? meta
          && lib.elem lib.categories.game.puzzle package.meta.categories
        );
      in
      result.success && result.value
    ) (lib.attrValues legacyPackages.x86_64-linux)
    ```

- [ ] Categories must have some sort of layering to them, so that users are not
      overwhelmed by having to look through hundreds of them. I suggest starting
      by adopting [Debian tags](https://packages.debian.org/about/debtags), with
      their two-layer facets + tags, based on the analysis in
      [Feature request: Optional package grouping](https://github.com/NixOS/nixpkgs/issues/556446).
- [ ] `package.meta.categories` is always set if `package.meta` exists, and is
      always a list. This simplifies the implementation: if the user doesn’t
      specify this attribute, it defaults to an empty list (assuming this is
      possible), meaning there’s no need for extra code to check for the
      existence of the `categories` attribute if `package.meta` exists.

## Maybe later

- Support categories within categories. This would complicate everything, and we
  already have two levels.
- Explain when adding/deleting/modifying facets/categories is appropriate. We
  can do this once the structure is more mature.
- Linking categories to equivalent ones in other package managers. We can do
  this if needed once categories are mature.

## Non-goals

- A fully generic categorisation system. This would result in a huge amount of
  theoretical and implementation work, and we need to be pragmatic enough to end
  up with a simple data structure and implementation.
- Reinventing the wheel. Other open source package managers have already done
  this, and we can leverage their work without needing to discuss the minutiae
  of which categories to include or how to structure them.
- Custom categories for individual packages (see below).
- Complete categorisation of all packages. We should spread the load of this as
  much as possible, by providing good documentation for how to do it, and
  encouraging people to add categories to packages they use/maintain. If we
  manage to provide a simple system for people to use, I would expect enough
  packages to be categorised within a release cycle to make filtering on
  categories useful.
- Something set in stone. Like everything else in nixpkgs, categories will
  probably change in major ways in the next few years, no matter how carefully
  we design it. Getting _something_ out there is more important than it being
  perfect.

# Plan

The phases below should be done in order.

## Alpha phase

Each of the tracks below should be able to run in parallel.

### Data gathering track (@l0b0)

- [x] Translate the existing
      [Debian tags](https://packages.debian.org/about/debtags) into Nix code
      imported into `lib/default.nix`, like

    ```nix
    # Facets and categories based on [Debian tags](https://packages.debian.org/about/debtags)
    {
      accessibility = {
        name = "Accessibility";
        description = "Accessibility support provided by the package";

        accessibilityViaAtSpi = {
          name = "Accessibility via AT-SPI";
          description = ''
            Applies to applications which are technically accessible through AT-SPI, e.g. the content of the interface can be accessed by the Orca screen reader.

            This does not imply that the application is convenient to use, only that the interface can be accessed.
          ''
        };
        # …
      };
      # …
    }
    ```

    Details:

    - The Debian tags system uses "facets" as a mandatory grouping structure
      around "tags". We will reuse this structure.
    - We rename the "tags" structure inside facets to "categories", per
      [RFC 146](https://github.com/NixOS/rfcs/blob/master/rfcs/0146-meta-categories.md).
    - Facet and category `name`s are plain strings, to avoid having to strip
      formatting for easy display in search interfaces.
    - Nesting the categories "properly", as in `lib.facets.x.categories.y` would
      be more explicit, but at the cost of longer code and not much clarity
      compared to `lib.categories.x.y`.
    - Avoid attrset keys which need to be quoted (if any) by normalising them.
    - Search interfaces should use the `name`s rather than the attrset keys to
      display names, since some keys will not be human readable.
    - Facet `name`s are typically the first paragraph of the Debian facet
      description.
    - Category `name`s are typically the first paragraph of the Debian tag
      description.
    - Only the category is directly referenced in `meta.category`, so category
      `name`s must be unique across all facets to make sure there's no
      ambiguity.
    - Facet and category `description`s are
      [CommonMark](https://commonmark.org/) Markdown strings (like
      `meta.longDescription`), so that they can include basic formatting.
      `description` is nullable, since the name might be enough to explain it.
    - Exclude the "TODO" tags, since any PR can suggest changes to categories.
    - Nested tags, such as "accessibility::**accessible-via:at-spi**", are rare,
      and are only implied by the string format. For now we should flatten them
      to simplify the implementation.
    - Attrset keys are ordered alphabetically according to the `en_US.utf8`
      locale.
    - Use your own judgement whether some of the facets/tags are never going to
      be relevant for nixpkgs, such as anything Debian-specific.

- [x] Include the [relevant license](https://www.debian.org/license) in the file
      with the categories.

Notes:

- [First draft of categories](https://github.com/NixOS/nixpkgs/compare/master...l0b0:nixpkgs:rfc-146-categorisation) for review/cherry-picking.

### Documentation track (@l0b0)

- [ ] Add a section to the nixpkgs reference documentation:
    - Reference the [Debian tags](https://packages.debian.org/about/debtags)
      origin, while making it clear that the Debian tags are a _starting point_,
      and not a source of truth - we will modify the structure as we go along
      for our own needs. Also mention that we have renamed their concept of
      "tags" to "categories", as per RFC 146.
    - Packages may _not_ create their own custom categories within the package.
      That is, they can't set
      `meta.categories = [{name = "my stuff"; description = "My own category";}]`,
      but must instead modify the existing categories attrset if they want to
      add one.
- [ ] Add a release note entry, including:
    - A reference to RFC 146 for anyone who wants details about the background.
    - Where to find the existing categories.
    - How to set categories for a package.
    - Encouraging people to add PRs with categories for packages they
      use/maintain.
    - A warning that the facets and categories themselves should be left
      unchanged until categories have been stabilised. If no category is
      appropriate, package maintainers should feel free to add them in separate
      PRs, with caveats:
        - The PR shouldn't be merged until the feature is stable, to avoid extra
          work for the categorisation group.
        - The underlying structure may change in the meantime, so they may have
          to resolve conflicts.

### Implementation track (@tr3foil)

- [ ] Add an empty list of categories to all existing packages' metadata.
      Ideally by defaulting this field value, if possible, so that we only need
      to modify a single place rather than every single package `meta`.
- [ ] Enforce "no custom categories" at ~~build~~ eval time. That is, every entry in
      `meta.categories` must be in the concatenation of all `lib.categories.*.*`
      values.
- [ ] Tag some of the most popular packages as a starting point.

## Beta phase

Once all of the above is in place, we should try to make sure it is as useful as
possible. This could all be done in parallel:

- [ ] Reach out to teams to ask them if they would be willing to categorise
      their own packages. We should make sure to mention that categorisation
      doesn’t need to be complete; for example, the GIS team could simply
      categorise everything they deal with as GIS software in the first PR.
- [ ] Implement various heuristics to categorise packages based on existing
      metadata, like name and description. Create small PRs to make sure
      packages with few users don’t accidentally get wrongly categorised because
      nobody reviewing the PR knows it.
- [ ] Reach out to the search team to ask whether the current structure lends
      itself to implementing extra filtering, or whether they would like to see
      any architectural changes.
- [ ] Add more documentation as needed:
    - Guidelines for modifying the facets/categories.

## Stable phase

- [ ] Disband the categorisation team.

**Comment by tr3foil** [comment]
> Enforce "no custom categories" at build time. That is, every entry in `meta.categories` must be in the concatenation of all `lib.categories.*.*` values.

I believe we want that enforced at eval time.

In fact, since `meta` isn't passed to the builder, I don't think it *can* be enforced at build time.

**Comment by jopejoe1** [comment]
> - [ ] Add an empty list of categories to all existing packages' metadata.
Ideally by defaulting this field value, if possible, so that we only need
to modify a single place rather than every single package `meta`.
> - [ ]  Enforce "no custom categories" at build time. That is, every entry in
`meta.categories` must be in the concatenation of all `lib.categories.*.*`
values.

Those can be added to [`./pkgs/stdenv/generic/check-meta.nixcheck-meta.nix`](https://github.com/NixOS/nixpkgs/blob/5b6a3bfd9b2ac2e6144b422189e884284de3d1f0/pkgs/stdenv/generic/check-meta.nix#L362), and a list of enums of the categories should work for the type.

**Comment by l0b0** [comment]
> > Enforce "no custom categories" at build time. That is, every entry in `meta.categories` must be in the concatenation of all `lib.categories.*.*` values.
> 
> I believe we want that enforced at eval time.
> 
> In fact, since `meta` isn't passed to the builder, I don't think it _can_ be enforced at build time.

Sorry, yes, I meant eval time. Updated

**Comment by llakala** [comment]
> I believe we want that enforced at eval time.

For eval performance for users, we should only do this in CI, and gate on config.checkMeta.

**Comment by tr3foil** [comment]
5 min initial implementation of the `meta.categories` interface: https://github.com/tr3foil/nixpkgs/tree/categories

I wanted to run the checks, but `nix-build ci -A eval.baseline` immediately used all my memory and swap... I'll have to figure out how to run the hydra eval checks on individual packages. I use a very test heavy workflow, I need to know if I'm doing it right at every step!

**Comment by tr3foil** [comment]
> > I believe we want that enforced at eval time.
> 
> For eval performance for users, we should only do this in CI, and gate on config.checkMeta.

I believe that `checkMeta` is already only run in ci, so no need to gate out categories specifically. If I'm wrong please correct me.

**Comment by tr3foil** [comment]
I've created a draft pr with my and @l0b0's work: https://github.com/NixOS/nixpkgs/pull/559645

**Comment by tr3foil** [comment]
I believe I have the implementation correct now but I would appreciate more eyes and testing!
